crates/accelerate/src/sabre/route.rs

```diff
@@ -328,7 +329,32 @@ impl<'a, 'b> RoutingState<'a, 'b> {
             current_swaps.push([shortest_path[end], shortest_path[end - 1]]);
         }
         current_swaps.iter().for_each(|&swap| self.apply_swap(swap));
-        closest_node
+
+        // If we apply a single swap it could be that we route 2 nodes; that is a setup like
+        //  A - B - A - B
+        // and we swap the middle two qubits. This cannot happen if we apply 2 or more swaps.
+        if current_swaps.len() > 1 {
+            smallvec![closest_node]
+        } else {
+            // check if the closest node has neighbors that are now routable -- for that we get
+            // the other physical qubit that was swapped and check whether the node on it
+            // is now routable
+            let mut possible_other_qubit = current_swaps[0]
+                .iter()
+                // check if other nodes are in the front layer that are connected by this swap
+                .filter_map(|&swap_qubit| self.front_layer.qubits()[swap_qubit.index()])
+                // remove the closest_node, which we know we already routed
+                .filter(|(node_index, _other_qubit)| *node_index != closest_node)
+                .map(|(_node_index, other_qubit)| other_qubit);
```

Comment:
Couldn't this just be a single filter map. Something like:

```suggestion
                .filter_map(|&swap_qubit| {
                    self.front_layer.qubits()[swap_qubit.index()]).map(|(node_index, other_qubit)| {
                        // remove the closest_node, which we know we already routed
                        if node_index != closest_node {
                            Some(other_qubit)
                        } else {
                            None
                        }
                    })
                });
```
Or even:

```suggestion
                .filter_map(|&swap_qubit| self.front_layer.qubits()[swap_qubit.index()])
                // remove the closest_node, which we know we already routed
                .filter_map(|(node_index, other_qubit)| if *node_index != closest_node {
                    Some(other_qubit)
                 } else {
                    None
                 })
```

Not that this makes a big difference. I'm surprised clippy isn't complaining about the filter followed by a map though.

Reply:
if you're really going hard on fancy methods, you can even do `(*node_index != closest_node).then_some(other_qubit)` haha.